File: Popcorn/popcorn.lua
-- starts

local filemask = "img%04d.tga"
local pool = 4 -- working threads pool size
local nx, ny = 1600, 1200 -- image dimension
local sensitivity = 0.02

-- xn+1 = xn - h sin(yn + tan(k yn))
-- yn+1 = yn - h sin(xn + tan(k xn))

local geomlst = {
--	{ h,       k,       },
	{ 0.05,    2.0,     },
	{ 0.05,    3.0,     },
	{ 0.05,    4.0,     },
	{ 0.05,    5.0,     },
}
local g = popcorn.new() -- make popcorn geometry structure
g.frames, g.iters = 10000, 20000 -- iterations parameters
g.minx, g.maxx = -8.0, 8.0
g.miny = g.minx * ny / nx
g.maxy = g.maxx * ny / nx

color.setpair(color.create(1.0, 0, 0), color.create(0, 0, 1.0))

local img = image.create(nx, ny)
for i, p in ipairs(geomlst) do
	print(string.format("h=%g k=%g", p[1], p[2]))
	g.h, g.k = p[1], p[2]

	img:clear()
	local dur = g:render(pool, img, "gradient") -- call multithreaded render
	print("\r"..dur.."s            ")

	local filename = string.format(filemask, i)
	io.write("writing...")
	img:writetga(filename, sensitivity)
	print("\r"..filename.."      ")
end

-- the end.

File: Popcorn/popcorn.cpp
/*
xn+1 = xn - h sin(yn + tan(k yn))
yn+1 = yn - h sin(xn + tan(k xn))
*/

#include "stdafx.h"

#include "popcorn.h"

using namespace attractors;

color hue1(1.0, 0, 0), hue2(0, 0, 1.0);

void popcorn::render(
	int quote, int pool,
	image& img,
	color::filter hue,
	std::function<void()> notify) const {

	color* bmp = img.data();
	int nx = img.nx, ny = img.ny;
	int fpc = frames / 100;

	// Make unique random sequence
	std::mt19937_64 prng(std::chrono::high_resolution_clock::now().time_since_epoch().count() + quote);
	std::uniform_real_distribution<number> distribution(0.0, 1.0);

	number dX = nx / (maxX - minX), dY = ny / (maxY - minY);

	int i1 = quote * frames / pool, i2 = (quote + 1) * frames / pool;
	for (int i = i1; i < i2; i++) {
		const number p = static_cast<number>(i) / frames;
		const color curcol = hue(p);

		number x = distribution(prng);
		number y = distribution(prng);
		number xn, yn;


		for (int j = 0; j < iters; j++) {
			xn = x - h * sin(y + tan(k * y));
			yn = y - h * sin(x + tan(k * x));
			x = xn;
			y = yn;

			int xi = static_cast<int>((x - minX) * dX);
			int yi = static_cast<int>((y - minY) * dY);
			if (xi >= 0 && xi < nx && yi >= 0 && yi < ny) {
				// there is no significant threads access to one data point at one time
				bmp[xi + yi * nx] += curcol;
			}
		}

		if (!(i % fpc)) {
			notify();
		}
	}
}

// The End.
